### PR TITLE
add free text search

### LINKED ISSUE #8: Improve free text search
## Acceptance Criteria
- [x] I can search for a guide using free text search
    - including external guides from API depends on #31 
- [x] When I search, the results make sense
- [x] I can search resources based on the current language
- [x] If I change language, the search is re-run
- [x] There is some indication of what has been searched for (e.g. text at the top of the page)
- [x] Search results are ordered based on how close of a match they are
- [x] Search implementation follows accessibility best practices

## Implementation Notes

- [x] use [fuse](https://fusejs.io/) search library based on this [comparison](https://fiatjaf.github.io/search-comparison/) test
- [x] search result are displayed using one of these methods
    -  Existing teaser list is filtered based on results
    - A pop up element appears based on a filtered teaser list
    - a combination of the above

- we want to include Actions in this search, so this needs doing first: #31 

## Elasticlunr
[http://elasticlunr.com/](http://elasticlunr.com/)

 from the getting started …

```jsx
var index = elasticlunr(function () {
    this.addField('title');
    this.addField('body');
    this.setRef('id');
});

var doc1 = {
    "id": 1,
    "title": "Oracle released its latest database Oracle 12g",
    "body": "Yestaday Oracle has released its new database Oracle 12g, this would make more money for this company and lead to a nice profit report of annual year."
}

var doc2 = {
    "id": 2,
    "title": "Oracle released its profit report of 2015",
    "body": "As expected, Oracle released its profit report of 2015, during the good sales of database and hardware, Oracle's profit of 2015 reached 12.5 Billion."
}

index.addDoc(doc1);
index.addDoc(doc2);
```

If all the json has uniform fields this will probably work nicely

- [https://github.com/kyrim/get-your-nutrients](https://github.com/kyrim/get-your-nutrients) has elasticlunr as dependency
    - json passed to model and elastic search port set up here https://github.com/kyrim/get-your-nutrients/blob/master/client/js/app.js 

## some elm packages that do search

- [https://package.elm-lang.org/packages/jxxcarlson/elm-text-search/latest/](https://package.elm-lang.org/packages/jxxcarlson/elm-text-search/latest/)
- [https://package.elm-lang.org/packages/jxxcarlson/elm-search/latest/](https://package.elm-lang.org/packages/jxxcarlson/elm-search/latest/)

- [https://github.com/elm/package.elm-lang.org/](https://github.com/elm/package.elm-lang.org/) has some sort of search on the side and on the homepage

## Fuse (probably the better option?)
https://fusejs.io/

Looks to be more actively maintained than elasticlunr

## Search-index
https://github.com/fergiemcdowall/search-index

### Ports
- https://guide.elm-lang.org/interop/ports.html

### Web component
- [https://guide.elm-lang.org/interop/custom_elements.html](https://guide.elm-lang.org/interop/custom_elements.html) 
- https://htmlwithsuperpowers.netlify.app/get-started/

### Examples
- [https://github.com/elm-community/js-integration-examples](https://github.com/elm-community/js-integration-examples)

### PR DESCRIPTION
**Pull request by aaaaargZombies**
Fixes #8 

## Description

- A web component takes in a `guideTeaserList` of local and remote guides/actions
- It returns a filtered list based on user input (produced by fuse.js) and updates `model.search`
- The guides page will display the contents of `model.search` instead of the full list of content if its not an empty list.
- The search input only appears on the `/guides` route.
- I've added a label to the input
- It doesn't allow you to search stories (it was a mistake to add this to AC)

### Possible improvements / future issues

- It does not alternate result based on language as discussed with Katja that can be added later.
- It does not re-order the list based on how closely it matches.
- design input for for label would be nice
- Allow searching of guides from all pages (more design input would help).

@geeksforsocialchange/developers

### CODE REVIEW

**Review by katjam — COMMENTED** [review]
This is looking pretty good.

File: src/Main.elm [hunk 1]
```diff
@@ -71,6 +71,7 @@ init flags url key =
             }
       , content = Page.Shared.Data.contentDictDecoder flags
       , language = English
+      , search = []
       , externalActions = Loading
```
[the remark this answers is not in the document]
Comment (aaaaargZombies) [reply src/Main.elm:75]: it's done in the init function

https://github.com/geeksforsocialchange/teamwilder/blob/115e7df82b56749c56cc07cfd101f0f04ed84973/src/Main.elm#L77

File: src/js/search.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  updateSearch() {
+    this.searchResult = this.fuse
+      .search(this._input.value)
+      .map((el) => el.item);
+    // because this code runs outside Elms event system we need to give
+    // Elm a bit of space to update the DOM when the attribute changes before
+    // updating. Perhaps there is some way to express this better in Elm.
+    //
```
[the remark this answers is not in the document]
Comment (katjam) [reply src/js/search.js:49]: Yeah this needs to run through Elm ports probably - or be internally implemented, but we can deal with that later.

[the remark this answers is not in the document]
Comment (aaaaargZombies) [reply src/js/search.js:49]: FYI it was fine without the setTimeout when just firing from user input but once it was being triggered by the attribute update the elm message seemed to "get stuck" and would then only fire when another elm message was triggered. sorry if that was confusing, it was a bit confusing....

**Pull request merged**